Zotero/Scenes/Detail/PDF/ViewModels/PDFReaderActionHandler.swift

```diff
@@ -1737,74 +1752,68 @@ final class PDFReaderActionHandler: ViewModelActionHandler, BackgroundDbProcessi
-            }
-
-        case .PSPDFAnnotationsAdded:
-            guard let annotations = notification.object as? [PSPDFKit.Annotation] else { return }
-            self.add(annotations: annotations, selectFirst: false, in: viewModel)
-
-        case .PSPDFAnnotationsRemoved:
-            guard let annotations = notification.object as? [PSPDFKit.Annotation] else { return }
-            self.remove(annotations: annotations, in: viewModel)
-
-        default: break
-        }
-
-        self.update(viewModel: viewModel) { state in
-            state.pdfNotification = notification
-        }
-    }
-
-    private func isNotificationFromDocument(_ notification: Notification, viewModel: ViewModel<PDFReaderActionHandler>) -> Bool {
-        if let annotation = notification.object as? PSPDFKit.Annotation {
-            return annotation.document == viewModel.state.document
-        }
-        if let annotations = notification.object as? [PSPDFKit.Annotation], let annotation = annotations.first {
-            return annotation.document == viewModel.state.document
+    private func observeDocument(in viewModel: ViewModel<PDFReaderActionHandler>) {
+        let nextBlock: (Notification) -> Void = { [weak self, weak viewModel] notification in
+            guard let self, let viewModel else { return }
+            processAnnotationObserving(handler: self, notification: notification, viewModel: viewModel)
         }
-        return false
-    }
 
-    private func observeDocument(in viewModel: ViewModel<PDFReaderActionHandler>) {
         NotificationCenter.default.rx
             .notification(.PSPDFAnnotationChanged)
             // Debounce these notifications because FreeTextAnnotation rotation change spams these annotations in milliseconds and it looks bad in sidebar while it's also unnecessary cpu burden
             .debounce(.milliseconds(100), scheduler: MainScheduler.instance)
-            .subscribe(onNext: { [weak self, weak viewModel] notification in
-                guard let self, let viewModel else { return }
-                self.processAnnotationObserving(notification: notification, viewModel: viewModel)
-            })
-            .disposed(by: self.pdfDisposeBag)
+            .subscribe(onNext: nextBlock)
+            .disposed(by: pdfDisposeBag)
 
         NotificationCenter.default.rx
             .notification(.PSPDFAnnotationsAdded)
             .observe(on: MainScheduler.instance)
-            .subscribe(onNext: { [weak self, weak viewModel] notification in
-                guard let self, let viewModel else { return }
-                self.processAnnotationObserving(notification: notification, viewModel: viewModel)
-            })
-            .disposed(by: self.pdfDisposeBag)
+            // Delay annotations added notifications, so they have completely recorded their undo command.
+            .delay(.microseconds(250), scheduler: MainScheduler.instance)
```

Comment:
This makes the worry above a bit worse :). What if the delay changes and we're removing wrong annotations?

Reply:
Check https://github.com/zotero/zotero-ios/pull/934/commits/22f4389ed5c9d0d55e0f245b90900d7b76efbb22 I've set the delay to `0`, and it still works. This is what we actually want, dispatch to the next main run loop iteration, so that the recording block of the `AnnotationStateManager` is finished.

Reply:
Removing this debounce (or well setting it to 0 milliseconds) defeats the comment's purpose a bit. Though text annotations are not enabled yet, so we can maybe keep it like this for now, but at least add a TODO there.

I'd probably also reach to PSDPFKit and explain our situation, maybe they could open the recording API a bit more so that we can search through the queue and remove only some steps manually.

Reply:
My bad, I changed the wrong debounce :(
Let me fix it and test again.

Reply:
Check https://github.com/zotero/zotero-ios/pull/934/commits/22546ac6f7301f41043768d30b9eef62a5755033